### DISCUSSION TITLE
Nested ItemsControls?

### DISCUSSION THREAD

**Discussion by andyabarnard** [AvaloniaUI/Avalonia#7328] [Q&A] [answered]
I have two classes, let's call them Person and Item.  One Person can own many or zero Items, each Item is owned by exactly one Person.

What I want to do is display a list of Person names, and below that, the names of the Items that Person owns.  Like this:

Person 1
    Item 1
    Item 2
Person 2
    Item 3
Person 4
    Item 4
    Item 5

I've got it more or less figured out, minus actually displaying the items.  My idea was to have a holder class that consists of a Person's name and a List of Items so I could bind that holder class to my ItemsControl, but I don't know how to actually display the contents of that List via Binding, if it's even possible.  I tried `Text="{Binding name} + {Binding items}"` but that doesn't do anything at all, beyond displaying the name.  Can I do this, or do I need to somehow nest ItemsControls?

Here's what my holder class looks like:

```
public class PersonAndItems
    {
        public string name { get; set; }
        
        public List<Item> items { get; set; }

        public PersonAndItems(string name, List<Item> items)
        {
            this.name= name;
            this.items = items;
        }
    }
```

EDIT: I've poked around a bit and found that placing this:

```
<StackPanel>
  <TextBlock TextAlignment="Center" Text="{Binding name}"/>
  <TextBlock TextAlignment="Center" Text="{Binding items[0].name}"/>
</StackPanel>
```

Inside the <DataTemplate> gets me _extremely_ close to what I want.  The only problem is I'd have to dynamically generate TextBlocks based on the size of the List of Items, because obviously just displaying the first item's name isn't quite enough.  If that's doable, then problem solved.  I know how to do this via code-behind, but everybody hates code-behind for reasons I don't fully understand so I'm trying to avoid it.  This should a pure MVVM solution, if possible.

**Comment by hendrikmennen** [comment, marked as answer]
You should have a look at Treeview.
Here is a good tutorial:
https://www.nequalsonelifestyle.com/2019/06/18/avalonia-treeview-tutorial/

**Reply by andyabarnard** [reply]
This looks to be exactly what I wanted.  My project just so happened to already be set up exactly the same way as this tutorial, too, so it took no time at all to implement.  Thank you.

**Comment by ahopper** [comment]
Nesting ItemsControls is an entirely normal thing to do, you just put the inner ItemsControl in the ItemTemplate for the outer one.

**Comment by wieslawsoltes** [comment]
Example of nesting https://github.com/zkSNACKs/WalletWasabi/blob/2e3a93dc25ab9fbe67dbfa487a054494fdebe006/WalletWasabi.Fluent/Views/Search/SearchPageView.axaml#L29-L76

```XAML
      <ItemsControl HorizontalAlignment="Left"
                    Items="{Binding SearchResults}"
                    Margin="6,0,20,0"
                    ClipToBounds="False">
        <ItemsControl.ItemsPanel>
          <ItemsPanelTemplate>
            <WrapPanel Orientation="Vertical" Margin="-8 0" />
          </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>
        <ItemsControl.ItemTemplate>
          <DataTemplate x:DataType="search:SearchResult">
            <DockPanel Margin="0,0,0,30">
              <Panel Margin="8 0"  DockPanel.Dock="Top">
                <Label Classes="h6 bold"
                       Margin="-6,0,0,0"
                       DataContext="{Binding Category}"
                       Content="{Binding Title}" />
              </Panel>
              <ItemsControl HorizontalAlignment="Left"
                            Items="{Binding Items}"
                            ClipToBounds="False">
                <ItemsControl.ItemsPanel>
                  <ItemsPanelTemplate>
                    <WrapPanel Orientation="Horizontal" />
                  </ItemsPanelTemplate>
                </ItemsControl.ItemsPanel>
                <ItemsControl.ItemTemplate>
                  <DataTemplate>
                    <Button Classes="option" Command="{Binding OpenCommand}" Height="115" Width="178" Margin="4">
                      <Panel>
                        <PathIcon Width="20" Height="20" Opacity="0.8" VerticalAlignment="Top" HorizontalAlignment="Left" Margin="12,12,0,0">
                          <PathIcon.Data>
                            <Binding Path="IconName" Converter="{x:Static conv:NavBarIconConverter.Instance}" />
                          </PathIcon.Data>
                        </PathIcon>
                        <TextBlock Text="{Binding Title}" Classes="h7 bold" VerticalAlignment="Top" HorizontalAlignment="Left" Margin="48 14 0 0" />
                        <Panel Margin="12,46,12,12">
                          <TextBlock Text="{Binding Caption}" Classes="h8" TextWrapping="Wrap" VerticalAlignment="Stretch" HorizontalAlignment="Stretch" Opacity="0.6" />
                        </Panel>
                      </Panel>
                    </Button>
                  </DataTemplate>
                </ItemsControl.ItemTemplate>
              </ItemsControl>
            </DockPanel>
          </DataTemplate>
        </ItemsControl.ItemTemplate>
      </ItemsControl>
```

**Reply by andyabarnard** [reply]
I'm new to Avalonia (and WPF/XAML in general) so I'm not fully comprehending what I'm seeing here, but this is definitely something I'll need to learn for the near future.  I'll keep this in mind for when it's learning time.  Thank you.
